### ISSUE TITLE
Install FOGBackup to fog path

### ISSUE THREAD

**Issue by jdorel** [FOGProject/fogproject#314]
Currently, if you want to use FOGBackup, you need to use it from the folder you downloaded (so no consistant path for other utilities). Or you can try to extract it (no the simplest thing because of the dependency on `lib/common`).

Either of those methods prevents easy updates.

So could the installer install `FOGBackup` to the fog path ? (Maybe a code generator could be used to include `lib/common` in the FOGBackup script to make without external dependencies)

**Comment by Sebastian-Roth** [comment]
@jdorel Again, sorry for the long delay.

I am wondering if we should just try to get rid of the dependency on `lib/common/utils.sh` as a path to go?!

The first two things that I found are the function call `handleError` and the variable `$webdirdest` that we'd definitely need to "take over" from the included script. Do you see anything else?

**Comment by jdorel** [comment]
I agree that getting rid of the dependency would be better. However I think the workload to get rid of them is an order of magnitude higher than the one to just modify the installer.

As I also think that modifying the installer is necessary to provide the utility scripts in the installation folder, rather than in the downloaded folder, I think modifying the installer should be the way to go first.

PS: I can't seem to find where the `osid` variable is set, although it appears to be extensively used ?

**Comment by Sebastian-Roth** [comment]
@jdorel said:
> However I think the workload to get rid of them is an order of magnitude higher than the one to just modify the installer.

Sorry about that but I am still not exactly sure I understand you right here. Are you saying that we should install `FOGBackup.sh` (plus `lib/common/...` stuff) into /opt/fogproject or some other path? In which situation does this make things easier?

`osid` is being set in function `displayOSChoices` - [see here](https://github.com/FOGProject/fogproject/blob/master/lib/common/functions.sh#L797). But it uses the variable `strSuggestedOS` which is derived from calling `lib/common/input.sh`. The more you dig through the scripts the more you will find that it's all a bit of a mess. The whole installer script grew over a period of several years while the developers (different people) tried to keep up with changes in the supported Linux systems as well as trying to support upgrades from older versions (both FOG and OS) and so on. There are so many checks in the scripts to make things work for all the different possible situations that I am sometimes close to a head ace when I try to get my head around the code. Feel free to ask if you think there is something wrong somewhere.

**Comment by jdorel** [comment]
> Sorry about that but I am still not exactly sure I understand you right here. Are you saying that we should install `FOGBackup.sh` (plus `lib/common/...` stuff) into /opt/fogproject or some other path? In which situation does this make things easier?

Yes, this would make the location of the utility scripts persistent across updates. And because the `lib/common/...` sourcing is relative, you can just keep the current folder structure and copy it to the installation path (`/opt/fogproject`)

**Comment by Sebastian-Roth** [comment]
@jdorel Sorry for the delay. I just don't seem to find the time to work on this. It's not forgotten but will take more time.

**Comment by mastacontrola** [comment]
I know it's been a very long time for this, but wouldn't it just be easier to download the installer to /opt/fog?

```
git clone https://github.com/fogproject/fogproject.git /opt/fog
```

Do actions from /opt/fog always?  At least for those using the FOGBackup.sh path and all?

Sure you'd have to remember it's in the utils, but I think this was how it was designed and why /opt/fog/.fogsettings became a thing.

**Comment by mastacontrola** [comment]
Done on `working-1.6` — `666f9385c`.

`installUtilities()` copies `utils/` and `lib/` into `$fogprogramdir`, which is
@jdorel's original proposal: keep the trees in their existing relative positions
and the `../../lib/common/...` sourcing keeps working untouched. `lib/` has to
come along because the utils source `lib/common/utils.sh`, which sources
`lib/common/functions.sh`. Both trees are removed before copying, so a file
dropped in a later release does not survive as a stale copy.

@Sebastian-Roth's alternative — severing the `lib/common` dependency — is still
the tidier end state, but @jdorel was right that it is an order of magnitude
more work, and it is not needed to give the utils a stable path.

One thing neither comment caught: a consistent path is not much use if you still
have to `cd` into it first, and you did. `.` resolves a relative path against
the **caller's** cwd, not the script's own location, so
`. ../../lib/common/utils.sh` only ever worked from inside the util's own
directory. `FOGBackup.sh`, `fogupdater.sh` and `lib/common/utils.sh` now resolve
against `BASH_SOURCE`, so this works from anywhere:

    /opt/fog/utils/FOGBackup/FOGBackup.sh -B /path/to/backups/

@mastacontrola — re: cloning the repo to `/opt/fog` instead: that is now the
same path without the clone, and without `.fogsettings`, `snapins/` and the CA
sharing a directory with a git working tree.

Verified on a Debian 12 container: install places all four utils and the five
`lib/` trees, `FOGBackup.sh` runs by absolute path from `/` and reaches its real
work, `--uninstall --dry-run` lists both new paths, and `--uninstall` removes
them while leaving the CA and the database intact.

`dev-branch` is not ported.

### Separate breakage this surfaced

`FOGBackup.sh` cannot complete on 1.6 wherever it lives. Its `backupDB` step
fetches `management/export.php?type=sql`, which is not in the 1.6 tree — the
request falls through to the API and gets a 403. `backupDB` runs unconditionally
and `handleError`s out, so the script dies before reports, snapins or images, and
no flag skips it.

Not fixed here: pointing it somewhere real means deciding how a shell script
authenticates to the web tier, which is not a call to make in passing. Worth its
own issue.

**Comment by mastacontrola** [comment]
Correction to my previous comment, and FOGBackup now completes — `11eab0225`.

I said fixing the database step meant deciding how a shell script authenticates to the web tier. It does not, and the reason I thought so was a wrong assumption about `maintenance/backup_db.php`. That endpoint is not missing from the tree — `bin/installfog.sh` deletes the whole `maintenance` directory at the end of every run. It is an unauthenticated same-machine DB dump, so it is install scaffolding by design, and that is why it answers during an install and 403s afterwards.

So FOGBackup cannot use it, and keeping it installed to suit this script would trade a permanent unauthenticated dump endpoint for convenience. But there was never an auth question: **FOGBackup runs on the server and `.fogsettings` already carries the database credentials**, so it dumps directly with `mysqldump`. Taking the web tier out of the path also means a backup still works when the web tier is broken, which is when you most want one.

Credentials go through a 0600 defaults file rather than the command line, where any local user could read them out of `ps` while the dump runs.

Also added `-D/--no-database`, since `backupDB` ran unconditionally — a storage node has no database of its own and could never get past it to back up its images.

Verified against a live 1.6 install with `maintenance/` already deleted:

```
 * Backing up database.........................................Done
 * Backing up reports..........................................Done
 * Backing up snapins..........................................Done
 * Backing up images...........................................Done
```

108KB dump carrying all 60 tables, plus reports, snapins (including the CA) and images, empty error log.

### One thing this turned up

The installer's own `backupDB` had the matching defect and was worse, because it is the pre-upgrade backup. It checked the status of whichever command ran last: `curl -skf ... | jq` makes that **jq's** status, and jq exits 0 on empty input having written a zero-byte file. An upgrade whose backup failed printed:

```
 * Backing up database.........................................Done
```

over an empty dump. Both halves of the pipeline are now checked, along with the result being non-empty and not the literal `null`, and a fresh install reports `Skipped` rather than `Done` for a step that never ran.

**Comment by mastacontrola** [comment]
Ported to `dev-branch` (1.5.x) as d3a7079ef + 717a3c7e0.

The utilities install to `$fogprogramdir/utils` and `$fogprogramdir/lib` on both the fresh-install and upgrade paths, and the `. ../../lib/common/utils.sh` sourcing now resolves via `BASH_SOURCE` instead of the caller's cwd — that was the reason the scripts only worked if you `cd`'d into their own directory first.

**FOGBackup was broken on 1.5 too, for a different reason than on 1.6.** 1.6 has no `management/export.php` at all. 1.5 does — but it is not a database endpoint. It calls `checkAuthAndCSRF()` and then exports whatever report is sitting in `$_SESSION['foglastreport']`. An unauthenticated `wget` satisfies neither check:

```
$ wget --post-data="nojson=1" -O /tmp/orig.sql "http://<server>//fog//management/export.php?type=sql"
HTTP request sent, awaiting response... 401 Unauthorized
  size: 0 bytes
```

So the script died on its first step and never reached reports, snapins or images. Both branches now dump directly with `mysqldump` using the credentials already in `.fogsettings` — no web tier in the path, which also means a backup still works when the web tier is broken.

Verified running by absolute path from `/` on a 1.5 install:

```
 * Started backup at: 08/01/26  02:49:46 PM
 * Backing up database.........................................Done
 * Backing up reports..........................................Done
 * Backing up snapins..........................................Done
 * Backing up images...........................................Done
 * Completed backup at: 08/01/26  02:49:46 PM

/tmp/bk/20260801_0/mysql/fog.sql  97591 bytes, 51 tables
```

Also on both branches: `-D/--no-database` (a storage node has no database of its own, and the unconditional `backupDB` meant it could never reach its images), a missing `dots()` label on the reports step, a zero-byte dump now counting as failure, and the installer's own pre-upgrade `backupDB` no longer reporting "Failed" for a step that never ran on a fresh install — it says "Skipped".

`lib/common/uninstall.sh` does not exist on `dev-branch`, so the working-1.6 change adding the two new trees to the uninstall listing has no counterpart here.

**Comment by mastacontrola** [comment]
Closing — both branches. `666f9385c`+`11eab0225` on working-1.6, `d3a7079ef`+`717a3c7e0` on dev-branch.

@jdorel's proposal as filed: `utils/` and `lib/` are installed to `$fogprogramdir` with their relative positions intact, so the existing sourcing keeps working and the utilities survive updates. Plus the thing that would have undermined it — the sourcing resolved against the caller's cwd, so a stable path was no use until it resolved against `BASH_SOURCE` instead.

@Sebastian-Roth's alternative of severing the `lib/common` dependency outright is still the tidier end state and is deliberately **not** done here. It is a larger piece of work and it is not what this issue asked for; if it is still wanted it deserves its own issue rather than staying attached to a request that is satisfied.

FOGBackup also actually completes now on both branches, which it had not on either.
